### PR TITLE
New Adapter: Zeta Global SSP

### PR DESCRIPTION
**Pull request by abermanov-zeta**
Docs: https://github.com/prebid/prebid.github.io/pull/4464

### CODE REVIEW

File: adapters/zeta_global_ssp/zeta_global_ssp.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	"net/http"
+
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type ZetaGlobalSspAdapter struct {
```
Comment (Sonali-More-Xandr) [inline adapters/zeta_global_ssp/zeta_global_ssp.go:15]: nitpick - Can we rename it to `adapter`?

File: static/bidder-info/zeta_global_ssp.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+endpoint: https://ssp.disqus.com/bid/prebid-server?sid=GET_SID_FROM_ZETA&shortname=GET_SHORTNAME_FROM_ZETA
+endpointCompression: gzip
+maintainer:
+  email: DL-Zeta-SSP@zetaglobal.com
```
Comment (Sonali-More-Xandr) [inline static/bidder-info/zeta_global_ssp.yaml:4]: @abermanov-zeta I have sent a verification email to this address. Please reply to that email to complete the maintainer email verification process.

Comment (Sonali-More-Xandr) [reply static/bidder-info/zeta_global_ssp.yaml:4]: email verification completed!

Comment (abermanov-zeta) [reply static/bidder-info/zeta_global_ssp.yaml:4]: Thank you

File: adapters/zeta_global_ssp/zeta_global_ssp.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
```
[the remark this answers is not in the document]
Comment (abermanov-zeta) [reply adapters/zeta_global_ssp/zeta_global_ssp.go]: Refactored. Thank you.

File: adapters/zeta_global_ssp/zeta_global_ssp.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		Headers: headers,
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func (a *ZetaGlobalSspAdapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
```
[the remark this answers is not in the document]
Comment (abermanov-zeta) [reply adapters/zeta_global_ssp/zeta_global_ssp.go]: Refactored. Thank you.

File: static/bidder-info/zeta_global_ssp.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+gvlVendorID: 833
```
Comment (Sonali-More-Xandr) [inline static/bidder-info/zeta_global_ssp.yaml:5]: The gvlVendorID for zetaglobal is 469. Can you please check?

File: adapters/zeta_global_ssp/zeta_global_ssp.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner == nil && imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+			return mediaType
+		}
+	}
```
Comment (gargcreation1992) [inline adapters/zeta_global_ssp/zeta_global_ssp.go:82]: This is an anti-pattern. Ideally, bidder should be setting BidType in their response. `response.seatbid[i].bid.ext.prebid.type` is where your bidder code should set it and then use the set BidType for [adapters.TypedBid ](https://github.com/prebid/prebid-server/pull/2676/files#diff-8def29254a193e276b10174c522ad09cc57e6250501eec803f328126c665ec2fR63)

File: static/bidder-info/zeta_global_ssp.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      - banner
+      - video
+  site:
+    mediaTypes:
+      - banner
+      - video
+userSync:
+  redirect:
+    url: https://ssp.disqus.com/redirectuser?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}&r={{redirect_url}}&partner=GET_SHORTNAME_FROM_ZETA
+    uid-macro: 'BUYERUID'
```
Comment (gargcreation1992) [inline static/bidder-info/zeta_global_ssp.yaml:19]: This doesn't seem to be correct and breaks the prebid-server code to build. You might want to correct `{{gdpr}}` with `{{.GDPR}}` 
Please also be sure that your cookie sync url work as expected and sets the correct macro.
You might also wanna replace `uid-macro` with `userMacro` and define its value as `"${BUYERUID}"`.

**Review by gargcreation1992 — CHANGES_REQUESTED** [review]
requesting few changes.

File: adapters/zeta_global_ssp/zeta_global_ssp.go (hunk 6 above)
Comment (abermanov-zeta) [reply adapters/zeta_global_ssp/zeta_global_ssp.go:88]: Refactored. Thank you

File: static/bidder-info/zeta_global_ssp.yaml (hunk 7 above)
Comment (abermanov-zeta) [reply static/bidder-info/zeta_global_ssp.yaml]: Sorry, it was copy-pasted from Java adapter. Fixed.

File: adapters/zeta_global_ssp/zeta_global_ssp.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if httputil.IsResponseStatusCodeNoContent(responseData) {
+		return nil, nil
+	}
+
+	if err := httputil.CheckResponseStatusCodeForErrors(responseData); err != nil {
+		return nil, []error{err}
+	}
```
Comment (gargcreation1992) [inline adapters/zeta_global_ssp/zeta_global_ssp.go:54]: Please add test cases for StatusCodeNoContent and CheckResponseStatusCodeForErrors check.

File: adapters/zeta_global_ssp/zeta_global_ssp.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+		var bidExt openrtb_ext.ExtBid
+		err := json.Unmarshal(bid.Ext, &bidExt)
+		if err == nil && bidExt.Prebid != nil {
+			return openrtb_ext.ParseBidType(string(bidExt.Prebid.Type))
+		}
+	}
+
+	return "", &errortypes.BadServerResponse{
+		Message: fmt.Sprintf("Failed to parse impression \"%s\" mediatype", bid.ImpID),
+	}
```
Comment (gargcreation1992) [inline adapters/zeta_global_ssp/zeta_global_ssp.go:92]: Please add a test case for this error scenario.

File: adapters/zeta_global_ssp/zeta_global_ssp-test/exemplary/no-bid.json [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+          },
+          "device": {
+            "ip": "123.123.123.123"
+          },
+          "test": 1,
+          "tmax": 500
+        }
+      },
+      "mockResponse": {
+        "status": 200,
```
Comment (gargcreation1992) [inline adapters/zeta_global_ssp/zeta_global_ssp-test/exemplary/no-bid.json:84]: Since this is no bid test case, it should be 204 instead of 200

**Review by gargcreation1992 — COMMENTED** [review]
Requesting to add test cases for couple of scenarios. Also please fix the PR check failure.

File: adapters/zeta_global_ssp/zeta_global_ssp.go (hunk 8 above)
Comment (abermanov-zeta) [reply adapters/zeta_global_ssp/zeta_global_ssp.go:54]: Added. Thank you.

File: adapters/zeta_global_ssp/zeta_global_ssp.go (hunk 9 above)
Comment (abermanov-zeta) [reply adapters/zeta_global_ssp/zeta_global_ssp.go:92]: Added. Thank you.

File: adapters/zeta_global_ssp/zeta_global_ssp-test/exemplary/no-bid.json (hunk 10 above)
Comment (abermanov-zeta) [reply adapters/zeta_global_ssp/zeta_global_ssp-test/exemplary/no-bid.json]: Sure, thank you.

**Review by gargcreation1992 — APPROVED** [review]
lgtm!

**Pull request merged**